dali/pipeline/operator/builtin/external_source.h

```diff
@@ -509,6 +533,31 @@ class ExternalSource : public Operator<Backend>, virtual public BatchSizeProvide
                   " and the current type is ", batch.type_info().name(), "."));
     previous_dtype_ = batch.type();
 
+    auto input_ndim = batch.shape().sample_dim();
+    if (HasNdim()) {
+      DALI_ENFORCE(input_ndim == ndim_,
+                   make_string("ExternalSource expected data with ", ndim_, " dimensions and got ",
+                     input_ndim, " dimensions."));
+    } else if (ndim_ != -1) {
+      DALI_ENFORCE(input_ndim == ndim_,
+                   make_string("Dimensionality of the data fed to the external source has "
+                      "changed from previous iteration. Dimensionality in the previous "
+                      "iteration was ", ndim_, " and the current is ", input_ndim, "."));
+    }
+    ndim_ = input_ndim;
+
+    if (spec_.HasArgument("layout")) {
```

Comment:
This is inconsistent with the behavior of InferNDim, which treats empty layout the same way as none at all. Which way is better is disputable, but we should be consistent.
If we go with treating empty strings as no layout, then:
```suggestion
    if (!layout_.empty()) {
```

Reply:
done